### PR TITLE
implement features.transformUsername on v2 OIE

### PR DESCRIPTION
**Pull request by sherizada-okta**
## Description:
Transforms `identifier` using the `settings.transformUsername` function.

An example transformUsername that is set in .widgetrc.js:
```
// The callback function is passed two arguments:
// 1) username: The name entered by the user
// 2) operation: The type of operation the user is trying to perform:
//      - PRIMARY_AUTH
//      - FORGOT_PASSWORD
//      - UNLOCK_ACCOUNT
transformUsername: function (username, operation) {
  // This example will append the '@acme.com' domain if the user has
  // not entered it
  return username.indexOf('@acme.com') > -1
    ? username
    : username + '@acme.com';
}
```
https://github.com/okta/okta-signin-widget#username-and-password

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [x] Added unit tests?

- [x] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video:
Screenshot 2021-03-11 at 2 12 27 PM
Screenshot 2021-03-11 at 2 12 15 PM

### Issue:

- [OKTA-314098](https://oktainc.atlassian.net/browse/OKTA-314098)

### CODE REVIEW

File: src/v2/controllers/FormController.js [hunk 1]
```diff
@@ -165,6 +165,12 @@ export default Controller.extend({
       return;
     }
 
+    const identifier = model.get('identifier');
+    const transformUsername = this.settings.get('features.transformUsername');
+    if (_.isFunction(transformUsername) && identifier) {
```
Comment (pradeepdewda-okta) [inline src/v2/controllers/FormController.js:170]: what if the identifier is an email ? We should check with backend on this..

Comment (sherizada-okta) [reply src/v2/controllers/FormController.js:177]: It's up to the dev what they want to do with transformUsername. 
Here's an example from SIW readme:
```
// The callback function is passed two arguments:
// 1) username: The name entered by the user
// 2) operation: The type of operation the user is trying to perform:
//      - PRIMARY_AUTH
//      - FORGOT_PASSWORD
//      - UNLOCK_ACCOUNT
transformUsername: function (username, operation) {
  // This example will append the '@acme.com' domain if the user has
  // not entered it
  return username.indexOf('@acme.com') > -1
    ? username
    : username + '@acme.com';
}
```

**Comment by pradeepdewda-okta** [comment]
this looks good.. adding some tests ?

**Comment by sherizada-okta** [comment]
@pradeepdewda-okta I'm thinking now of moving this logic inside BaseModel.toJSON.

File: src/v2/controllers/FormController.js [hunk 2]
```diff
@@ -17,6 +17,7 @@ import { FORMS } from '../ion/RemediationConstants';
 import Util from '../../util/Util';
 import { clearTransactionMeta } from '../client/transactionMeta';
 
+
```
Comment (haishengwu-okta) [inline src/v2/controllers/FormController.js:20]: remove extra empty line

File: src/v2/view-builder/internals/BaseModel.js [hunk 3]
```diff
@@ -11,6 +11,13 @@
  */
 import { _, Model } from 'okta';
 
+const FORM_NAME_TO_OPERATION_MAP = {
+  'select-authenticator-unlock-account': 'UNLOCK_ACCOUNT',
+  'identify': 'PRIMARY_AUTH',
+  'select-identify': 'PRIMARY_AUTH',
```
Comment (haishengwu-okta) [inline src/v2/view-builder/internals/BaseModel.js:17]: Doesn't find what view represented by `select-identify`?
Looks like it's just an action to switch from register page to identify page?

File: src/v2/view-builder/internals/BaseModel.js [hunk 4]
```diff
@@ -80,6 +87,22 @@ const create = function (remediation = {}, optionUiSchemaConfig = {}) {
   const BaseModel = Model.extend({
     props,
     local,
+    toJSON: function () {
+      const modelJSON = Model.prototype.toJSON.apply(this, arguments);
+
+      const identifier = this.get('identifier');
+      if (identifier) {
```
Comment (haishengwu-okta) [inline src/v2/view-builder/internals/BaseModel.js:94]: shall it be `identifier && currentFormName is one of Object.keys(FORM_NAME_TO_OPERATION_MAP)`?

**Review by haishengwu-okta — COMMENTED** [review]
Do we have counter parter in v2 for `'signin/recovery-challenge'` which looks like also call transforUserName

File: src/v2/view-builder/internals/BaseModel.js [hunk 5]
```diff
@@ -11,6 +11,13 @@
  */
 import { _, Model } from 'okta';
 
+const FORM_NAME_TO_OPERATION_MAP = {
+  'select-authenticator-unlock-account': 'UNLOCK_ACCOUNT',
```
Comment (haishengwu-okta) [inline src/v2/view-builder/internals/BaseModel.js:15]: use constant from `Enum.js` ?

File: src/v2/view-builder/internals/BaseModel.js [hunk 6]
```diff
@@ -80,6 +87,22 @@ const create = function (remediation = {}, optionUiSchemaConfig = {}) {
   const BaseModel = Model.extend({
     props,
     local,
+    toJSON: function () {
```
Comment (haishengwu-okta) [inline src/v2/view-builder/internals/BaseModel.js:90]: I'm thinking another option is to call transformUserName at `FormController.handleSaveForm`, right before `idx.proceed`.
This option will work for `processCreds` which could be pure function and async function (return promise)

Comment (sherizada-okta) [reply src/v2/view-builder/internals/BaseModel.js:90]: That option is what I originally had - https://github.com/okta/okta-signin-widget/pull/1730/commits/c01f10a98d8939d557aee90045d7fc114caffed1

But I didn't find an existing way to test the FormController, and now I think its easier for testing to put it here and a more intuitive place for it.

processCreds would make sense in the FormController.

File: src/v2/view-builder/internals/BaseModel.js (hunk 4 above)
Comment (sherizada-okta) [reply src/v2/view-builder/internals/BaseModel.js:94]: That would make this setting bound to our upkeep of this Enum, which would be annoying for us and users.

**Comment by sherizada-okta** [comment]
> Do we have counter part in v2 for `'signin/recovery-challenge'` which looks like also call transformUserName

@haishengwu-okta I don't understand this question, can you rephrase? These changes are for v2.

File: src/v2/view-builder/internals/BaseModel.js (hunk 3 above)
Comment (sherizada-okta) [reply src/v2/view-builder/internals/BaseModel.js:17]: enroll-profile-new.json and enroll-profile.json have an identifier form value for it:
```
{
        "rel": [
          "create-form"
        ],
        "name": "select-identify",
        "href": "http://localhost:3000/idp/idx",
        "method": "POST",
        "accepts": "application/vnd.okta.v1+json",
        "value": [
          {
            "name": "identifier",
            "label": "identifier"
          },
          {
            "name": "stateHandle",
            "value": "01r2p5S9qaAjESMFuPzt7r3ZMcZZQ_vvS0Tzg56ajB",
            "visible": false
          }
        ]
      }

Comment (haishengwu-okta) [reply src/v2/view-builder/internals/BaseModel.js:17]: I didn't find a View represented for `select-identify` in ViewFactory hence the ask.

I may miss something or we need to double check if the mock is out of dated.

File: src/v2/view-builder/internals/BaseModel.js (hunk 6 above)
Comment (haishengwu-okta) [reply src/v2/view-builder/internals/BaseModel.js:90]: Hard to test seems not a good justification cause you'll need to test `processCreds` anyway? :D 
In test cafe, We could re-render ui with different set of options so that you'd be able to verify the identifier changed (in request body)

**Comment by haishengwu-okta** [comment]
Maybe we don't have `signin/recovery-challenge` in V2 (I don't quite remember). It's just question and worth to double check.

File: src/v2/view-builder/internals/BaseModel.js (hunk 6 above)
Comment (sherizada-okta) [reply src/v2/view-builder/internals/BaseModel.js:90]: > Hard to test seems not a good justification cause you'll need to test `processCreds` anyway? :D
heh I know, but I wanted to put off crossing that bridge until I really had to.

**Review by pradeepdewda-okta — APPROVED** [review]
💯

File: src/util/Enums.js [hunk 7]
```diff
@@ -79,4 +79,10 @@ export default {
   UNIVERSAL_LINK_CHALLENGE: 'UNIVERSAL_LINK',
 
   LAUNCH_AUTHENTICATOR_REMEDIATION_NAME: 'launch-authenticator',
+
+  FORM_NAME_TO_OPERATION_MAP: {
```
Comment (haishengwu-okta) [inline src/util/Enums.js:83]: this var probably shall go into `RemediationConstant` but uses `Enums.PRIMARY_AUTH`, etc.

File: src/v2/controllers/FormController.js [hunk 8]
```diff
@@ -165,6 +167,22 @@ export default Controller.extend({
       return;
     }
 
+    const identifier = model.get('identifier');
```
Comment (haishengwu-okta) [inline src/v2/controllers/FormController.js:170]: I'd rather to change json result instead of model itself.
```
const requestValue = model.toJSON();
if requestValue.identify then
  ...
idx.proceed(formName, requestValue);
```

**Review by haishengwu-okta — APPROVED** [review]
- I think testcafe is good enough without unit test.
- Move around the Enum variable and 🚀

**Comment by sherizada-okta** [comment]
> Maybe we don't have `signin/recovery-challenge` in V2 (I don't quite remember). It's just question and worth to double check.

The recovery flows that take an identifier are unlock account and forgot password.
